### PR TITLE
seal DefaultHttpContext and remove obsolete members

### PR DESCRIPTION
**Pull request by davidfowl**
Spoke to @DamianEdwards offline about this breaking change so putting it here in a PR to review. I'm sure I need to update other things as well but this is the baseline. Motivation for this change comes from this PR https://github.com/aspnet/AspNetCore/pull/6424

PS:  There are a class of other changes related to the DefaultHttpContext that we want to optimize:

https://github.com/aspnet/HttpAbstractions/pull/999

cc @benaadams

### CODE REVIEW

**Comment by JamesNK** [comment]
Question from someone not super familiar with the internals of Http:

What is the reason for choosing to make `DefaultHttpContext` sealed compared to a new class that is a implementation of DefaultHttpContext?

```cs
public sealed class ReusableHttpContext : DefaultHttpContext { }
```

There would be overhead ensuring changes to DefaultHttpContext are then made reusable in ReusableHttpContext, but it would avoid the breaking change. Sealing could break a lot of test code based around mocking, e.g. `Mock<DefaultHttpContext>`

**Comment by davidfowl** [comment]
> What is the reason for choosing to make DefaultHttpContext sealed compared to a new class that is a implementation of DefaultHttpContext?

More overhead, more fields, more virtual calls. The fields inside of `DefaultHttpContext` are typed as the base type instead of the derived type because of the "flexibility".

> There would be overhead ensuring changes to DefaultHttpContext are then made reusable in ReusableHttpContext, but it would avoid the breaking change. Sealing could break a lot of test code based around mocking, e.g. Mock<DefaultHttpContext>

Yes, that's the biggest risk in this change. Test code trying to override members of DefaultHttpContext without using features.

**Comment by halter73** [comment]
So you're planning on deleting ReusableHttpContext and going back to using DefaultHttpContext?

**Comment by davidfowl** [comment]
> So you're planning on deleting ReusableHttpContext and going back to using DefaultHttpContext?

Right, if this gets approval. That's the idea.

@JamesNK I could unseal the DefaultHttpContext to make a bit less breaking so that `Mock<DefaultHttpContext>` would still work but I'm not there yet.

**Comment by JamesNK** [comment]
> Yes, that's the biggest risk in this change. Test code trying to override members of DefaultHttpContext without using features.

Ok! A mitigation could be to have something like `TestableHttpContext` in an official testing package. It would provided default virtual implementations without being sealed. Developers would then replace `Mock<DefaultHttpContext>` in tests with TestableHttpContext upon updating to 3.0.

**Comment by davidfowl** [comment]
> Ok! A mitigation could be to have something like TestableHttpContext in an official testing package. It would provided default virtual implementations without being sealed. Developers would then replace Mock<DefaultHttpContext> in tests with TestableHttpContext upon updating to 3.0.

Yes, that's a decent mitigation as well.

**Comment by halter73** [comment]
> Ok! A mitigation could be to have something like TestableHttpContext in an official testing package. It would provided default virtual implementations without being sealed. Developers would then replace `Mock<DefaultHttpContext>` in tests with `TestableHttpContext` upon updating to 3.0.

If you're using Moq anyway, what's wrong with `Mock<HttpContext>`?

It's also a little unfortunate that DefaultHttpContext.(Un)Initialize() is still public now that it's no longer virtual. Is there any reason a developer should ever call these instead of letting the server handle the pooling?

**Comment by JamesNK** [comment]
@halter73 `Mock<HttpContext>` would be far more common.

The difference is `HttpContext` requires you to mock out all the behavior that you need, while `DefaultHttpContext` lets you take the defaults and override only what you care about.

Some results for mocking the default context, but less than I would have guessed: https://github.com/search?q=%22Mock%3CDefaultHttpContext%3E%22&type=Code

File: src/Http/Http/src/DefaultHttpContext.cs [hunk 1]
```diff
@@ -26,7 +26,7 @@ public sealed class DefaultHttpContext : HttpContext
 
         private DefaultHttpRequest _request;
         private DefaultHttpResponse _response;
```
Comment (halter73) [inline src/Http/Http/src/DefaultHttpContext.cs]: Nit: readonly

Comment (davidfowl) [reply src/Http/Http/src/DefaultHttpContext.cs]: Can't, they get re-initialized in Initialize

**Comment by halter73** [comment]
Not fast enough to beat the email :P

**Comment by davidfowl** [comment]
:trollface:

**Comment by benaadams** [comment]
Make the break 😍

Is a `DefaultHttpContext` not a CustomHttpContext 😉 

Alternative would be to make the changes to `DefaultHttpContext` (minus init/uninit); then inhert for a `PooledHttpContext` and seal that adding the init/uninit methods; just to avoid copypasta for Features if creating your own HttpContext

**Comment by rynowak** [comment]
> The difference is HttpContext requires you to mock out all the behavior that you need, while DefaultHttpContext lets you take the defaults and override only what you care about.

In N years of building MVC core, I haven't happened upon a situation where mocking the context is a better choice than newing up a `DefaultHttpContext`. Do we know of cases like this?

**Comment by davidfowl** [comment]
> In N years of building MVC core, I haven't happened upon a situation where mocking the context is a better choice than newing up a DefaultHttpContext. Do we know of cases like this?

Most of the tests I see, new up the DefaultHttpContext (it was actually intentional to make it POCO like and newable). A few purists mock the HttpContext and 10 people on github mock the DefaultHttpContext. I'm sure even less have a derived type outside of that last scenario besides @benaadams. If it's too big of a break, we can unseal the DefaultHttpContext but we'd likely be missing out on some JIT optimizations (though TBH I'm not 100% sure if there are other things preventing that from happening).

**Comment by davidfowl** [comment]
So I'm guessing @DamianEdwards is going to have to sign off on this one, or @Eilon

**Review by DamianEdwards — APPROVED** [review]
Breaking change approved at this stage given the data and discussed potential mitigations we could employ later on.

**Comment by Eilon** [comment]
Is this announcement-worthy?

**Comment by Tratcher** [comment]
@HaoK this overlaps with https://github.com/aspnet/AspNetCore/pull/4485

**Comment by davidfowl** [comment]
https://github.com/aspnet/Announcements/issues/338

**Comment by HaoK** [comment]
I can rebase my PR after

**Pull request merged**